data/transactions/logic/eval.go

```diff
@@ -2802,36 +2776,114 @@ func opAppDeleteGlobalState(cx *evalContext) {
 		return
 	}
 
-	err := cx.appDeleteGlobalKey(key)
+	err := cx.Ledger.DelGlobal(key)
 	if err != nil {
 		cx.err = err
 		return
 	}
 	cx.stack = cx.stack[:last]
 }
 
-func opAssetHoldingGet(cx *evalContext) {
-	last := len(cx.stack) - 1 // asset id
-	prev := last - 1          // account offset
+// We have a difficult naming problem here. In some opcodes, TEAL
+// allows (and used to require) ASAs and Apps to to be referenced by
+// their "index" in an app call txn's foeign-apps or foreign-assets
+// arrays.  That was a small integer, no more than 2 or so, and was
+// often called an "index".  But it was not a basics.AssetIndex or
+// basics.ApplicationIndex.
 
-	assetID := cx.stack[last].Uint
-	accountIdx := cx.stack[prev].Uint
-	fieldIdx := uint64(cx.program[cx.pc+1])
+func appReference(cx *evalContext, ref uint64, foreign bool) (basics.AppIndex, error) {
+	if cx.version >= directRefEnabledVersion {
+		if ref == 0 {
+			return cx.Ledger.ApplicationID(), nil
+		}
+		if ref <= uint64(len(cx.Txn.Txn.ForeignApps)) {
+			return basics.AppIndex(cx.Txn.Txn.ForeignApps[ref-1]), nil
+		}
+		for _, appID := range cx.Txn.Txn.ForeignApps {
+			if appID == basics.AppIndex(ref) {
+				return appID, nil
+			}
+		}
+		// It should be legal to use your own app id, which
+		// can't be in ForeignApps during creation, because it
+		// is unknown then.  But it can be discovered in the
+		// app code.  It's tempting to combine this with the
+		// == 0 test, above, but it must come after the check
+		// for being below len(ForeignApps)
+		if ref == uint64(cx.Ledger.ApplicationID()) {
+			return cx.Ledger.ApplicationID(), nil
+		}
+	} else {
+		// Old rules
+		if foreign {
+			// In old versions, a foreign reference must be an index in ForeignAssets or 0
+			if ref == 0 {
+				return cx.Ledger.ApplicationID(), nil
+			}
+			if ref <= uint64(len(cx.Txn.Txn.ForeignApps)) {
+				return basics.AppIndex(cx.Txn.Txn.ForeignApps[ref-1]), nil
+			}
+		} else {
+			// Otherwise it's direct
+			return basics.AppIndex(ref), nil
+		}
+	}
+	return basics.AppIndex(0), fmt.Errorf("invalid App reference %d", ref)
+}
+
+func asaReference(cx *evalContext, ref uint64, foreign bool) (basics.AssetIndex, error) {
+	if cx.version >= directRefEnabledVersion {
+		// In recent versions, accept either kind of ASA reference
+		if ref < uint64(len(cx.Txn.Txn.ForeignAssets)) {
+			return basics.AssetIndex(cx.Txn.Txn.ForeignAssets[ref]), nil
+		}
+		for _, assetID := range cx.Txn.Txn.ForeignAssets {
+			if assetID == basics.AssetIndex(ref) {
+				return assetID, nil
+			}
+		}
```

Comment:
`opAssetHoldingGet` calls `asaReference` with `foreign=false`. New code will fail to find it and fails, but old version would return `basics.AssetIndex(ref)`

The idea why asset id is not in `ForeignAssets` but account is in `Accounts` is - holdings are in balance records, so loading an entry from `Accounts` automatically fetches all asset holdings.

See comment:
```
data/transactions/application.go:       // ForeignAssets are asset IDs for assets whose AssetParams may be read
```